tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -0,0 +1,166 @@
+ */
+@Beta
+public final class LockFreeExponentiallyDecayingReservoir implements Reservoir {
+    private static final int DEFAULT_SIZE = 1028;
+    private static final double DEFAULT_ALPHA = 0.015;
+    private static final long DEFAULT_RESCALE_THRESHOLD = TimeUnit.HOURS.toNanos(1);
+
+    private static final AtomicReferenceFieldUpdater<LockFreeExponentiallyDecayingReservoir, State> stateUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(LockFreeExponentiallyDecayingReservoir.class, State.class, "state");
+
+    private final double alpha;
+    private final int size;
+    private final long rescaleThresholdNanos;
+    private final Clock clock;
+
+    private volatile State state;
+
+    private final class State {
+        private final long startTick;
+        private final AtomicLong count;
+        private final ConcurrentSkipListMap<Double, WeightedSample> values;
+
+        State(long startTick, AtomicLong count, ConcurrentSkipListMap<Double, WeightedSample> values) {
+            this.startTick = startTick;
+            this.values = values;
+            this.count = count;
+        }
+
+        private void update(long value, long timestampNanos) {
+            final double itemWeight = weight(timestampNanos - startTick);
+            final WeightedSample sample = new WeightedSample(value, itemWeight);
+            final double priority = itemWeight / ThreadLocalRandom.current().nextDouble();
+
+            final long newCount = count.incrementAndGet();
+            if (newCount <= size || values.isEmpty()) {
+                values.put(priority, sample);
+            } else {
+                Double first = values.firstKey();
+                if (first < priority && values.putIfAbsent(priority, sample) == null) {
+                    // Always remove an item
+                    while (values.remove(first) == null) {
+                        first = values.firstKey();
+                    }
+                }
+            }
+        }
+
+        State rescale(long newTick) {
+            long durationNanos = newTick - startTick;
+            double durationSeconds = durationNanos / 1_000_000_000D;
+            double scalingFactor = Math.exp(-alpha * durationSeconds);
+            AtomicLong newCount = new AtomicLong();
+            ConcurrentSkipListMap<Double, WeightedSample> newValues = new ConcurrentSkipListMap<>();
+            if (Double.compare(scalingFactor, 0) != 0) {
+                values.forEach((key, sample) -> {
+                    double newWeight = sample.weight * scalingFactor;
+                    if (Double.compare(newWeight, 0) == 0) {
+                        return;
+                    }
+                    WeightedSample newSample = new WeightedSample(sample.value, newWeight);
+                    if (newValues.put(key * scalingFactor, newSample) == null) {
+                        newCount.incrementAndGet();
```

Comment:
why increment the `newCount` variable on every forEach lambda call rather than just set it once at the end to `newValues.size()` like they did? https://github.com/dropwizard/metrics/blob/release/4.1.x/metrics-core/src/main/java/com/codahale/metrics/ExponentiallyDecayingReservoir.java#L184-L185

Reply:
Size isn't trivial, and we're already iterating through elements.

Reply:
boils down to a longAdder sum? 

I guess I don't know of the top of my head whether a longadder sum vs incrementing a non-atomic singleton long array vs incrementing an AtomicLong is quicker 🤷

Reply:
```suggestion
                        newCount++;
```

more a micro perf nit, but `newCount` could just be a `long` as we don't need the volatile/JMM side effects in this single thread until we create the new state, e.g. something like follows (and related suggested changes):
```
            long newCount = 0;
            ConcurrentSkipListMap<Double, WeightedSample> newValues = new ConcurrentSkipListMap<>();
            if (Double.compare(scalingFactor, 0) != 0) {
                values.forEach((key, sample) -> {
                    double newWeight = sample.weight * scalingFactor;
                    if (Double.compare(newWeight, 0) == 0) {
                        return;
                    }
                    WeightedSample newSample = new WeightedSample(sample.value, newWeight);
                    if (newValues.put(key * scalingFactor, newSample) == null) {
                        newCount++;
                    }
                });
            }
            return new State(newTick, new AtomicLong(newCount), newValues);
```

Reply:
We'd either have to allocate a final container for the non-volatile value, or replace forEach with an iterator. What do you think?